jetty-fcgi/fcgi-client/src/main/java/org/eclipse/jetty/fcgi/client/http/HttpConnectionOverFCGI.java

```diff
@@ -373,6 +374,12 @@ private Delegate(HttpDestination destination)
             super(destination);
         }
 
+        @Override
+        protected Iterator<HttpChannel> getHttpChannels()
+        {
+            return new IteratorWrapper<>(activeChannels.values().iterator());
```

Comment:
No, because `activeChannels.values().iterator()` has a type that is a subclass of `HttpChannel`, so the code won't compile.

This is a compile error because the caller obtains a "container" of types `HttpChannel` that actually contains `HttpChannelOverFCGI`, but the caller would be entitled to add to this "container" a different subclass, say `HttpChannelOverHTTP2`, which would break FCGI.

The fact that `Iterator` is a read-only data structure is unfortunately not taken into consideration -- I was surprised too, but that's just how generics work in Java, so the wrapper is required.

Reply:
Shame that the method you are overriding couldn't be `protected Iterator<? extends HttpChannel> getHttpChannels()` instead?

Reply:
@joakime returning wildcard types is a no-no in every book.